tests/audioworklet/audioworklet.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright 2015 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <pthread.h>
+#include <math.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <assert.h>
+
+#include <atomic>
+#include <mutex>
+
+#include <emscripten/emscripten.h>
+#include <emscripten/em_js.h>
+#include <emscripten/threading.h>
+
+static pthread_t workletThreadId = 0; // This is set on the main thread when the pthread is created
+static uint32_t sampleRate = 48000; 
+static float invSampleRate = 1.0f / sampleRate;
+
+///////////////////////////////////////////////////////////////////////////////
+// Section 1 - Test-only code
+//
+// This is just the testing code used to verify the pthread setup in the audio
+// worklet. This can be ignored if you're just wanting to see how to implement
+// native code audio worklets with Emscripten.
+///////////////////////////////////////////////////////////////////////////////
+
+std::atomic<pthread_t> pthreadIdFromWorklet(0); // This is set from the audio thread when the AudioWorkletNode is constructed
+std::mutex testMutex; // A test mutex to test the worklet futex implementation
+
+// Sends the current pthread id from the audio worklet node, signalling test success
+// This happens immediately in the worklet node constructor since 'process' is never
+// invoked without user interaction
+EMSCRIPTEN_KEEPALIVE extern "C" void signalTestSuccess() {
+  pthread_t threadId = pthread_self();
+  pthreadIdFromWorklet.store(threadId);
+
+  printf("Audio thread time: %f\n", emscripten_get_now());
+
+  // Put the worklet to sleep until the main thread sees the above store.
+  // NOTE: This is *really bad* and should never be done in an audio worklet, but
+  // is here purely to test the worklet futex implementation which is sometimes needed.
+  testMutex.lock(); 
+  testMutex.unlock();
+}
+
+// The main loop keeps going until 'pthreadIdFromWorklet' is sent from the AudioWorkletNode
+EM_BOOL mainLoop(double time, void* userdata) {
+  pthread_t threadId = pthreadIdFromWorklet.load();
+  if(threadId != 0) {
+    printf("Main thread time: %f\n", emscripten_get_now());
+
+    if(threadId == workletThreadId) {
+      printf("Sucesss! Got pthread id: %lu, expected %lu\n", threadId, workletThreadId);
+      #ifdef REPORT_RESULT
+        REPORT_RESULT(1);
+      #endif
+    } else {
+      printf("Failed! Got wrong pthread id: %lu, expected %lu\n", threadId, workletThreadId);
+      #ifdef REPORT_RESULT
+        REPORT_RESULT(-100);
+      #endif
+    }
+
+    testMutex.unlock();
+    return EM_FALSE;
+  }
+
+  return EM_TRUE;
+}
+
+EMSCRIPTEN_KEEPALIVE extern "C" void beforeUnload() {
+  // Always release the mutex on unload, otherwise the worklet can get stuck if we navigate away
+  testMutex.unlock();
```

Comment:
But won't the worklet get killed in this case?   Is it really possible to "leak" a worklet when you navigate away?     This is not true for normal pthreads/workers.

Reply:
This test is testing the worklet futex support so in `signalTestSuccess` the worklet tries to lock the mutex and it starts waiting for the main thread. If the main thread never makes it to the code that unlocks the mutex (e.g. you kill it before by navigating away) the worklet remains stuck in a loop and the tab becomes unresponsive. 

So it won't _leak_ the worklet, it'll try to clean it up and it'll become unresponsive :)

Workers don't do this, because they use `Atomics.Wait/Notify` to wait on a mutex which is non-blocking, but if the main thread was waiting on a mutex and the worker that's holding it crashes and doesn't unlock it then the same thing will happen (tab unresponsive)